tests/resolve-lock/resolve_lock.go

```diff
@@ -0,0 +1,493 @@
+}
+
+func (c *resolveLockClient) Start(ctx context.Context, cfg interface{}, clientNodes []types.ClientNode) error {
+	log.Infof("start to test...")
+	defer func() {
+		log.Infof("test end...")
+	}()
+
+	lastGreenGC := -1
+
+	for loopNum := 0; ; loopNum++ {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		err := c.generateLocks(ctx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		// Get a ts as the safe point so it's greater than any locks written by `generateLocks`
+		safePoint, err := c.getTs(ctx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		// Invoke GC with the safe point
+		id := fmt.Sprintf("gc-worker-%v", loopNum)
+		greenGCUsed, err := gcworker.RunDistributedGCJob(ctx, c.kv, c.pd, safePoint, id, 3, c.EnableGreenGC)
+		if err != nil {
+			log.Errorf("failed to run GC at safe point %v. This may be caused by nemesis or there is really something wrong.")
+		}
+
+		if greenGCUsed {
+			lastGreenGC = loopNum
+		}
+
+		if c.EnableGreenGC && loopNum-lastGreenGC > 10 {
+			return errors.New("green gc failed to run for over 10 times")
+		}
+
+		// Check there is no lock before safePoint
+		err = c.CheckData(ctx, safePoint)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+}
+
+// generateLocks sends Prewrite requests to TiKV to generate locks, without committing and rolling back.
+func (c *resolveLockClient) generateLocks(ctx context.Context) error {
+	log.Infof("Start generating lock")
+	errCh := make(chan error, c.GenerateLockConcurrency)
+	key := make([]byte, 0)
+	lastEndKey := make([]byte, 0)
+	currentRunningRegions := 0
+
+	for {
+		regions, _, err := c.pd.ScanRegions(ctx, key, []byte{}, 256)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		nextScanKey := regions[len(regions)-1].EndKey
+		nextScanKeyCopy := make([]byte, len(nextScanKey))
+		copy(nextScanKeyCopy, nextScanKey)
+
+		for _, region := range regions {
+			err = decodeRegionMetaKey(region)
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			if len(region.EndKey) > 0 && bytes.Compare(region.EndKey, lastEndKey) <= 0 {
+				continue
+			}
+			if bytes.Compare(region.StartKey, lastEndKey) < 0 {
+				region.StartKey = lastEndKey
+			}
+			lastEndKey = region.EndKey
+
+			for currentRunningRegions >= c.GenerateLockConcurrency {
+				select {
+				case <-ctx.Done():
+					return nil
+				case err = <-errCh:
+					if err != nil {
+						return errors.Trace(err)
+					}
+					currentRunningRegions--
+				}
+			}
+			currentRunningRegions++
+			go func() {
+				// Retry for several times to reduce the possibility of failing when the cluster is unstable
+				var err error
+				for retry := 0; retry < 10; retry++ {
+					err = c.singleLockRegion(ctx, region.StartKey, region.EndKey)
+					if err == nil {
+						errCh <- nil
+						return
+					}
+				}
+				errCh <- err
+			}()
+		}
+
+		key = nextScanKeyCopy
+		if len(key) == 0 {
+			break
+		}
+	}
+
+	for currentRunningRegions >= c.GenerateLockConcurrency {
```

Comment:
It's useless. Do you intend to leave some goroutines running during resolving locks?